toolchain/parser/parser.cpp

```diff
@@ -1658,4 +1663,75 @@ auto Parser::HandleVarFinishAsNoSemicolonState() -> void {
   HandleVarFinish(/*require_semicolon=*/false);
 }
 
+auto Parser::HandleInterfaceIntroducerState() -> void {
+  auto state = PopState();
+
+  if (!ConsumeAndAddLeafNodeIf(TokenKind::Identifier(),
+                               ParseNodeKind::DeclaredName())) {
+    CARBON_DIAGNOSTIC(ExpectedInterfaceName, Error,
+                      "Expected interface name after `interface` keyword.");
+    emitter_->Emit(*position_, ExpectedInterfaceName);
+    state.has_error = true;
+  }
+
+  bool has_open_curly_brace = true;
+
+  if (!PositionIs(TokenKind::OpenCurlyBrace())) {
+    CARBON_DIAGNOSTIC(ExpectedInterfaceOpenCurlyBrace, Error,
+                      "Expected `{{` to start interface definition.");
+    emitter_->Emit(*position_, ExpectedInterfaceOpenCurlyBrace);
+    state.has_error = true;
+
+    if (auto next_open_curly_brace = FindNextOf({TokenKind::OpenCurlyBrace()});
```

Comment:
I might've missed this before, but this is an undesirable recovery because the open curly could be at any later point in the file. e.g.:

```
interface Foo
class Bar { <- uses this open curly
}
```

For now, I think it'd be better to just call `SkipPastLikelyEnd(state.token);` and don't handle the open curly. The SkipPastLikelyEnd has a bit more logic to it which should help catch issues, and using it may also make it easier to find similar cases when we start trying to improve errors.